libsnark/zk_proof_systems/plonk/srs.tcc

```diff
@@ -106,12 +106,119 @@ usrs<ppT> plonk_usrs_derive_from_secret(
 
 template<typename ppT>
 srs<ppT> plonk_srs_derive_from_usrs(
-    const usrs<ppT> &usrs, const circuit_t<ppT> &circuit)
+    const usrs<ppT> &usrs,
+    const std::vector<std::vector<libff::Fr<ppT>>> gates_matrix,
+    const std::vector<size_t> wire_permutation,
+    const std::vector<size_t> PI_wire_indices)
 {
+    using Field = libff::Fr<ppT>;
+    const std::vector<std::vector<Field>> gates_matrix_transpose =
+        plonk_gates_matrix_transpose(gates_matrix);
+
+    // the number of gates is equal to the number of columns in the transposed
+    // gates matrix
+    size_t num_gates = gates_matrix_transpose[0].size();
+    // ensure that num_gates is not 0
+    assert(num_gates > 0);
+    // ensure num_gates is power of 2
+    assert((num_gates & (num_gates - 1)) == 0);
+
+    // the number of Q-polynomials (aka selector polynomials) is equal to the
+    // number of rows in the transposed gates matrix
+    size_t num_qpolys = gates_matrix_transpose.size();
+
+    // the constraints q_L, q_R, q_O, q_M, q_C and the
+    // witness w_L, w_R, w_O are represented as polynomials in the roots of
+    // unity e.g. f_{q_L}(omega_i) = q_L[i], 0\le{i}<8
+    std::shared_ptr<libfqfft::evaluation_domain<Field>> domain =
+        libfqfft::get_evaluation_domain<Field>(num_gates);
+
+    // compute the selector polynomials (q-polynomials) from the
+    // transposed gates matrix over the Lagrange basis q_poly = \sum_i
+    // q[i] * L[i] where q[i] is a coefficient (a scalar Field
+    // element) and L[i] is a polynomial with Field coefficients
+    std::vector<polynomial<Field>> Q_polys =
+        plonk_compute_selector_polynomials<Field>(
+            num_gates, num_qpolys, gates_matrix_transpose, domain);
+
+    // An explanation of the constants k1,k2 from [GWC19], Section 8,
+    // page 26 :
+    //
+    // "We explicitly define the multiplicative subgroup H as
+    // containing the n-th roots of unity in F_p , where w (omega) is
+    // a primitive n-th root of unity and a generator of H i.e: H =
+    // {1, w, ... , w^{n-1}}. We assume that the number of gates in a
+    // circuit is no more than n.
+
+    //
+    // For the moment k1,k2 are fixed to the test vector values for
+    // debug purpouses. TODO choose k1,k2 according to the
+    // requirements in [GWC19]
+    libff::Fr<ppT> k1 =
+        Field("706987411474581393682955260879121390206111740035659671471"
```

Comment:
Where do these values come from?

Reply:
These are from the test vector values, which are part of the `example` class. Shall I add this in the comments?

Reply:
Another important thing to mention is that these are specific to the BLS12-381 curve.
Don't need to say they came from a specific example, but we shoudl just state that these satisfy the requirements for BLS12-381.

Reply:
Fixed.